proof/verifier.go

```diff
@@ -180,71 +226,276 @@ func (p *Proof) verifySplitRootProof() error {
+			err := p.handleBasicExclusionProof(
+				&baseProof, p2trOutputs, commitVersions,
+			)
+			if err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		_, err := p.verifySTXOProofSet(
+			baseProof, assetMap, p2trOutputs, commitVersions, false,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = p.verifyRemainingOutputs(p2trOutputs)
+	if err != nil {
+		return nil, err
+	}
+
 	if len(commitVersions) == 0 {
 		return nil, nil
 	}
 
-	// All ExclusionProofs must have similar versions.
-	firstCommitVersion := maps.Values(commitVersions)[0]
-	for outputIdx, commitVersion := range commitVersions {
-		outputCommitVersion := commitVersion
-		if !commitment.IsSimilarTapCommitmentVersion(
-			&firstCommitVersion, &outputCommitVersion,
-		) {
+	// All proofs must have similar versions.
+	firstVersion := maps.Values(commitVersions)[0][0]
+	return p.verifySTXOVersions(firstVersion, commitVersions)
+}
+
+// handleBasicExclusionProof handles a basic (non-STXO) exclusion proof.
+func (p *Proof) handleBasicExclusionProof(baseProof *TaprootProof,
+	p2trOutputs map[uint32]fn.Set[asset.SerializedKey],
+	commitVersions map[uint32][]commitment.TapCommitmentVersion) error {
+
+	derivedCommitment, err := verifyTaprootProof(
+		&p.AnchorTx, baseProof, &p.Asset, false,
+	)
+	if err != nil {
+		return fmt.Errorf("error verifying STXO proof: %w", err)
+	}
+
+	outputIdx := baseProof.OutputIndex
+	delete(p2trOutputs, baseProof.OutputIndex)
+
+	// Store commitment version if Taproot Asset commitment present.
+	if derivedCommitment != nil {
+		commitVersions[outputIdx] = append(
+			commitVersions[outputIdx],
+			derivedCommitment.Version,
+		)
+	}
+
+	return nil
+}
+
+// verifySTXOProofSet verifies a set of STXO proofs.
+func (p *Proof) verifySTXOProofSet(baseProof TaprootProof,
+	assetMap map[asset.SerializedKey]*asset.Asset,
+	p2trOutputs map[uint32]fn.Set[asset.SerializedKey],
+	commitVersions map[uint32][]commitment.TapCommitmentVersion,
+	inclusion bool,
+) (*commitment.TapCommitment, error) {
+
+	var commitment *commitment.TapCommitment
+	for key := range baseProof.CommitmentProof.STXOProofs {
+		stxoProof := baseProof.CommitmentProof.STXOProofs[key]
+		stxoAsset := assetMap[key]
+		stxoCombinedProof := p.makeSTXOProof(baseProof, &stxoProof)
+
+		var err error
+		commitment, err = verifyTaprootProof(
+			&p.AnchorTx, &stxoCombinedProof, stxoAsset, inclusion,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error verifying STXO "+
+				"proof: %w", err)
+		}
+
+		outIdx := stxoCombinedProof.OutputIndex
+		delete(p2trOutputs[outIdx], key)
+		if len(p2trOutputs[outIdx]) == 0 {
+			delete(p2trOutputs, outIdx)
+		}
+
+		commitVersions[outIdx] = append(
+			commitVersions[outIdx], commitment.Version,
+		)
+	}
+
+	return commitment, nil
+}
+
+// makeSTXOProof creates a new proof for an STXO reusing the base proof while
+// replacing commitmentProof bu the stxoProof.
+func (p *Proof) makeSTXOProof(baseProof TaprootProof,
```

Comment:
This can be a pure function, looks like we don't use `p` anywhere. Then we can export it and also use it in the itest in the previous commit. Though it feels a bit weird that we have to fake a whole `TaprootProof` just to verify the STXO proof. But let's not refactor that right now, just thinking aloud.

Reply:
Don't get me started on the `TaprootProof` construction. I don't like it _at all_.

Reply:
Yeah, me neither... Probably something we can refactor at some point.